frontend/src/app/collaboration/editor/editor.component.ts

```diff
@@ -0,0 +1,238 @@
+                parser: 'java',
+                plugins: [prettierPluginJava, prettierPluginEstree], // Add necessary plugins
+            });
+
+            this.view.dispatch({
+                changes: {
+                    from: 0,
+                    to: this.view.state.doc.length,
+                    insert: await formattedCode,
+                },
+            });
+
+            this.view.focus();
+        } catch (e) {
+            console.error('Error formatting code:', e);
+            this.messageService.add({ severity: 'error', summary: 'Formatting Error' });
+        }
+    }
+
+    setProvider() {
+        const randomIndex = Math.floor(Math.random() * usercolors.length);
+
+        this.wsProvider.awareness.setLocalStateField('user', {
+            name: this.authService.userValue?.username,
+            color: usercolors[randomIndex].color,
+            colorLight: usercolors[randomIndex].light,
+        });
+    }
+
+    setEditorState() {
+        const undoManager = this.undoManager;
+        const myExt: Extension = [
+            basicSetup,
+            java(),
+            this.customTheme,
+            oneDark,
+            yCollab(this.yeditorText, this.wsProvider.awareness, { undoManager }),
+        ];
+
+        this.state = EditorState.create({
+            doc: this.yeditorText.toString(),
+            extensions: myExt,
+        });
+    }
+
+    setEditorView() {
+        const editorElement = this.editor.nativeElement;
+        const state = this.state;
+        this.view = new EditorView({
+            state,
+            parent: editorElement,
+        });
+    }
+
+    setTheme() {
+        this.customTheme = EditorView.theme(
+            {
+                '&': {
+                    backgroundColor: 'var(--surface-section)',
+                },
+                '.cm-gutters': {
+                    backgroundColor: 'var(--surface-section)',
+                },
+            },
+            { dark: true },
+        );
+    }
+
+    setCursorPosition() {
+        // set new cursor position
+        const cursorPosition = this.state.doc.line(1).from;
+
+        this.view.dispatch({
+            selection: {
+                anchor: cursorPosition,
+                head: cursorPosition,
+            },
+        });
+
+        this.view.focus();
+    }
+
+    onSubmitDialogClose() {
+        this.messageService.add({
+            severity: 'error',
+            summary: 'Fail',
+            detail: 'Submission failed: Not all participants agreed. Please try again.',
+        });
+        this.isSubmit = false;
+        this.isInitiator = false;
+    }
+
+    onSuccess() {
+        this.messageService.add({
+            severity: 'success',
+            summary: 'Success',
+            detail: 'You have successfully submitted!',
+        });
+        this.isSubmit = false;
+        this.isInitiator = false;
+    }
+
+    forfeit() {
+        this.isForfeitClick = true;
+    }
```

Comment:
After one user has forfeited, the other user still gets the dialogue "Forfeiting would result in your peer working alone."